### PR TITLE
[expo-notifications][iOS] Swift conversion 3: Scheduling, NotificationBuilder, NotificationCenterDelegate, Presentation

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Migrate scheduler, notification builder, and presentation module to Swift

# How

- Rewrite scheduler module in Swift
- Rewrite presentation module in Swift
- Create singleton to receive notification center events and allow modules to register as delegates for those events
- Replace Objective C dictionary categories with a Swift implementation using Records

# Test Plan

- CI should pass, including SwiftLint checks
Behavior of notification test app on real device should not change

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by douglowder** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/33253?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#33253** <a href="https://app.graphite.dev/github/pr/expo/expo/33253?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/expo/expo/33253?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#32612** <a href="https://app.graphite.dev/github/pr/expo/expo/32612?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#32069** <a href="https://app.graphite.dev/github/pr/expo/expo/32069?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationObjcWrapper.m [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2024-present 650 Industries. All rights reserved.
+
+#import "EXNotificationObjcWrapper.h"
+
+@implementation EXNotificationObjcWrapper
+
+/**
+ Wrapper that allows Swift code to execute Objective-C code that could throw an NSException, which is not catchable by Swift.
+ */
++ (BOOL)tryExecute:(nonnull void(NS_NOESCAPE^)(void))tryBlock error:(__autoreleasing NSError * _Nullable * _Nullable)error {
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationObjcWrapper.m:10]: Maybe you can use this: https://github.com/expo/expo/blob/4c9c62752347568b5174cd537d3de15d03f58b65/packages/expo-modules-core/ios/Legacy/EXUtilities.m#L218-L228

File: packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  func openSettings(_ notification: UNNotification?)
+}
+
+public protocol NotificationRegistrationSuccessDelegate: AnyObject {
+  func didRegister(_ deviceToken: String)
+}
+
+public protocol NotificationRegistrationFailureDelegate: AnyObject {
+  func didFailRegistration(_ error: Error)
+}
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift:24]: Do we need so many protocols? I believe we can have one and provide default (empty) implementations in the protocol extension.

File: packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ */
+@objc(EXNotificationCenterManager)
+public class NotificationCenterManager: NSObject,
+  UNUserNotificationCenterDelegate,
+  NotificationRegistrationFailureDelegate,
+  NotificationRegistrationSuccessDelegate {
+  @objc
+  public static let shared = NotificationCenterManager()
+
+  var delegates: [AnyObject] = []
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift:38]: If we have just one protocol for delegates (see a comment above), I guess we won't need an array of delegates.

File: packages/expo-notifications/ios/EXNotifications/Notifications/Presenting/PresentationModule.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
+//  Copyright © 2024 650 Industries. All rights reserved.
+
+import ExpoModulesCore
+import UIKit
+import MachO
+
+public class PresentationModule: Module, NotificationPresentationDelegate {
+  var presentedNotifications: Set<String> = []
+
+  public func willPresent(_ notification: UNNotification, completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
```
Comment (tsapeta) [inline packages/expo-notifications/ios/EXNotifications/Notifications/Presenting/PresentationModule.swift:10]: Let's keep the `definition` function on top of the class

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationObjcWrapper.m (hunk 1 above)
Comment (douglowder) [reply packages/expo-notifications/ios/EXNotifications/Notifications/EXNotificationObjcWrapper.m:10]: Thanks! I'll try that

File: packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift (hunk 2 above)
Comment (douglowder) [reply packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift:24]: I was attempting to duplicate the previous Objective-C code that would see if the methods were implemented -- I agree empty implementations will be simpler

File: packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift (hunk 3 above)
Comment (douglowder) [reply packages/expo-notifications/ios/EXNotifications/Notifications/NotificationCenterManager.swift:38]: There will be multiple modules registering as delegates, so I think we will still need the array

**Review by tsapeta — APPROVED** [review]

**Pull request merged**